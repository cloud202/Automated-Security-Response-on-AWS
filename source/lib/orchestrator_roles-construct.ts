#!/usr/bin/env node
/*****************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import {
    Stack,
    Construct,
    ArnFormat } from '@aws-cdk/core';
import { 
    PolicyStatement, 
    Effect, 
    Role, 
    PolicyDocument, 
    ArnPrincipal,
    ServicePrincipal,
    CompositePrincipal, 
    CfnRole 
} from '@aws-cdk/aws-iam';

export interface OrchRoleProps {
    solutionId: string;
    adminAccountId: string;
    adminRoleName: string;
}

export class OrchestratorMemberRole extends Construct {
  constructor(scope: Construct, id: string, props: OrchRoleProps) {
    super(scope, id);
    const RESOURCE_PREFIX = props.solutionId.replace(/^DEV-/,''); // prefix on every resource name
    const stack = Stack.of(this);
    const memberPolicy = new PolicyDocument();

    /**
     * @description Cross-account permissions for Orchestration role
     * @type {PolicyStatement}
     */
    const iamPerms = new PolicyStatement();
    iamPerms.addActions(
        "iam:PassRole",
        "iam:GetRole"
    )
    iamPerms.effect = Effect.ALLOW
    iamPerms.addResources(
        `arn:${stack.partition}:iam::${stack.account}:role/${RESOURCE_PREFIX}-*`
    );
    memberPolicy.addStatements(iamPerms)

    const ssmRWPerms = new PolicyStatement()
    ssmRWPerms.addActions(
        "ssm:StartAutomationExecution"
    )
    ssmRWPerms.addResources(
        stack.formatArn({
            service: 'ssm',
            region: '*',
            resource: 'document',
            resourceName: 'SHARR-*',
            arnFormat: ArnFormat.SLASH_RESOURCE_NAME
        }),
        stack.formatArn({
            service: 'ssm',
            region: '*',
            resource: 'automation-definition',
            resourceName: '*',
            arnFormat: ArnFormat.SLASH_RESOURCE_NAME
        }),
        stack.formatArn({
            service: 'ssm',
            region: '*',
            resource: 'automation-definition',
            account:'',
            resourceName: '*',
            arnFormat: ArnFormat.SLASH_RESOURCE_NAME
        }),
        stack.formatArn({
            service: 'ssm',
            region: '*',
            resource: 'automation-execution',
            resourceName: '*',
            arnFormat: ArnFormat.SLASH_RESOURCE_NAME
        })
    );
    memberPolicy.addStatements(ssmRWPerms)

    memberPolicy.addStatements(
        // The actions in your policy do not support resource-level permissions and require you to choose All resources
        new PolicyStatement({
            actions: [
                'ssm:DescribeAutomationExecutions',
                'ssm:GetAutomationExecution'
            ],
            resources: [ '*' ],
            effect: Effect.ALLOW
        }),
        new PolicyStatement({
            actions: [
                'ssm:DescribeDocument'
            ],
            resources: [ `arn:${stack.partition}:ssm:*:*:document/*` ],
            effect: Effect.ALLOW
        }),
        new PolicyStatement({
            actions: [
                'ssm:GetParameters',
                'ssm:GetParameter'
            ],
            resources: [ `arn:${stack.partition}:ssm:*:*:parameter/Solutions/SO0111/*` ],
            effect: Effect.ALLOW
        }),
        new PolicyStatement({
            actions: [
                "config:DescribeConfigRules"
            ],
            resources: [ "*" ],
            effect: Effect.ALLOW
        })
    )

    const sechubPerms = new PolicyStatement();
    sechubPerms.addActions("cloudwatch:PutMetricData")
    sechubPerms.addActions("securityhub:BatchUpdateFindings")
    sechubPerms.effect = Effect.ALLOW
    sechubPerms.addResources("*")

    memberPolicy.addStatements(sechubPerms)

    let principalPolicyStatement = new PolicyStatement();

    principalPolicyStatement.addActions("sts:AssumeRole");
    principalPolicyStatement.effect = Effect.ALLOW;

    let roleprincipal = new ArnPrincipal(
        `arn:${stack.partition}:iam::${props.adminAccountId}:role/${props.adminRoleName}`
    );

    let principals = new CompositePrincipal(roleprincipal);
    principals.addToPolicy(principalPolicyStatement);

    let serviceprincipal = new ServicePrincipal('ssm.amazonaws.com')
    principals.addPrincipals(serviceprincipal);

    let memberRole = new Role(this, 'MemberAccountRole', {
        assumedBy: principals,
        inlinePolicies: {
            'member_orchestrator': memberPolicy
        },
        roleName: `${RESOURCE_PREFIX}-SHARR-Orchestrator-Member`
    });

    const memberRoleResource = memberRole.node.findChild('Resource') as CfnRole;

    memberRoleResource.cfnOptions.metadata = {
        cfn_nag: {
            rules_to_suppress: [{
                id: 'W11',
                reason: 'Resource * is required due to the administrative nature of the solution.'
            },{
                id: 'W28',
                reason: 'Static names chosen intentionally to provide integration in cross-account permissions'
            }]
        }
    }
  }
}