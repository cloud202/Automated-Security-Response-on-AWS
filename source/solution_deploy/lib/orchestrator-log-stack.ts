#!/usr/bin/env node
/*****************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.        *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import * as cdk from '@aws-cdk/core';
import { LogGroup, CfnLogGroup, RetentionDays } from '@aws-cdk/aws-logs';
import { Key } from '@aws-cdk/aws-kms';
 
export interface OrchLogStackProps extends cdk.StackProps {
    description: string;
    logGroupName: string;
    solutionId: string;
}
  
export class OrchLogStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props: OrchLogStackProps) {
  
    super(scope, id, props);
    const stack = cdk.Stack.of(this);

    const reuseOrchLogGroup = new cdk.CfnParameter(this, 'Reuse Log Group', {
        type: "String",
        description: `Reuse existing Orchestrator Log Group? Choose "yes" if the log group already exists, else "no"`,
        default: "no",
        allowedValues: ["yes", "no"]
    })
    reuseOrchLogGroup.overrideLogicalId(`ReuseOrchestratorLogGroup`)

    const kmsKeyArn = new cdk.CfnParameter(this, 'KMS Key Arn', {
        type: "String",
        description: `ARN of the KMS key to use to encrypt log data.`,
    })
    kmsKeyArn.overrideLogicalId(`KmsKeyArn`)
     
    /**********************
     * Encrypted log group
     */
    // As of March 2021, CWLogs encryption is not yet supported in GovCloud
    // Choose based on partition

    const kmsKey = Key.fromKeyArn(this, "KmsKey", kmsKeyArn.valueAsString)
    const orchestratorLogGroupEncrypted: LogGroup = new LogGroup(this, 'Orchestrator-Logs-Encrypted', {
        logGroupName: props.logGroupName,
        removalPolicy: cdk.RemovalPolicy.RETAIN,
        retention: RetentionDays.ONE_YEAR,
        encryptionKey: kmsKey
    });

    /************************
     * Unencrypted log group
     */
    const orchestratorLogGroupNOTEncrypted: LogGroup = new LogGroup(this, 'Orchestrator-Logs', {
        logGroupName: props.logGroupName,
        removalPolicy: cdk.RemovalPolicy.RETAIN,
        retention: RetentionDays.ONE_YEAR
    });
 
    /*******************
     *  Conditions
     */
    const isNotGovCloud = new cdk.CfnCondition(this, "isNotGovCloud", {
        expression: cdk.Fn.conditionNot(cdk.Fn.conditionEquals(stack.partition, "aws-us-gov"))
    });
    {
        let childToMod = orchestratorLogGroupEncrypted.node.defaultChild as CfnLogGroup;
        childToMod.cfnOptions.condition = new cdk.CfnCondition(this, "Encrypted Log Group", {
            expression: cdk.Fn.conditionAnd(
                isNotGovCloud,
                cdk.Fn.conditionEquals(reuseOrchLogGroup.valueAsString, "no")
            )
        })
    }

    {
        let childToMod = orchestratorLogGroupNOTEncrypted.node.defaultChild as CfnLogGroup;
        childToMod.cfnOptions.condition = new cdk.CfnCondition(this, "Unencrypted Log Group", {
            expression: cdk.Fn.conditionAnd(
                cdk.Fn.conditionNot(isNotGovCloud),
                cdk.Fn.conditionEquals(reuseOrchLogGroup.valueAsString, "no")
            )
        })
        childToMod.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W84',
                    reason: 'KmsKeyId is not supported in GovCloud.'
                }]
            }
        }
    }
  }
} 