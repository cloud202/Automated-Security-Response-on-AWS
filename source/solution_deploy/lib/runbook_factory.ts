#!/usr/bin/env node
/******************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.        *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import { IssmPlaybookProps, RemediationRunbookProps } from '../../lib/ssmplaybook';
import * as cdk_nag from 'cdk-nag';
import * as lambda from '@aws-cdk/aws-lambda';
import * as s3 from '@aws-cdk/aws-s3';
import * as iam from '@aws-cdk/aws-iam';
import * as cdk from '@aws-cdk/core';
import * as fs from 'fs';

export interface RunbookFactoryProps {
  solutionId: string;
  runtimePython: lambda.Runtime;
  solutionDistBucket: string;
  solutionTMN: string;
  solutionVersion: string;
  region: string;
  partition: string;
};

export class RunbookFactory extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string, props: RunbookFactoryProps) {
    super(scope, id);

    const RESOURCE_PREFIX = props.solutionId.replace(/^DEV-/, '');

    const policy = new iam.Policy(this, 'Policy', {
      policyName: RESOURCE_PREFIX + '-SHARR_Runbook_Provider_Policy',
      statements: [
        new iam.PolicyStatement({
          actions: [
            'cloudwatch:PutMetricData'
          ],
          resources: ['*']
        }),
        new iam.PolicyStatement({
          actions: [
            'logs:CreateLogGroup',
            'logs:CreateLogStream',
            'logs:PutLogEvents'
          ],
          resources: ['*']
        }),
        new iam.PolicyStatement({
          actions: [
            'ssm:CreateDocument',
            'ssm:UpdateDocument',
            'ssm:UpdateDocumentDefaultVersion',
            'ssm:ListDocumentVersions',
            'ssm:DeleteDocument'
          ],
          resources: ['*']
        })
      ]
    });

    const cfnPolicy = policy.node.defaultChild as iam.CfnPolicy;
    cfnPolicy.cfnOptions.metadata = {
      cfn_nag: {
        rules_to_suppress: [
          {
            id: 'W12',
            reason: 'Resource * is required in order to manage arbitrary SSM documents'
          }
        ]
      }
    };

    cdk_nag.NagSuppressions.addResourceSuppressions(policy, [
      {id: 'AwsSolutions-IAM5', reason: 'Resource * is required in order to manage arbitrary SSM documents'}
    ]);

    const role = new iam.Role(this, 'Role', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      description: 'Lambda role to allow creation of updatable SSM documents'
    });

    role.attachInlinePolicy(policy);

    const SolutionsBucket = s3.Bucket.fromBucketAttributes(this, 'SolutionsBucket', {
      bucketName: props.solutionDistBucket + '-' + props.region
    });

    const memberLambdaLayer = new lambda.LayerVersion(this, 'MemberLambdaLayer', {
      compatibleRuntimes: [props.runtimePython],
      description: 'SO0111 SHARR Common functions used by the solution member stack',
      license: 'https://www.apache.org/licenses/LICENSE-2.0',
      code: lambda.Code.fromBucket(
          SolutionsBucket,
          props.solutionTMN + '/' + props.solutionVersion + '/lambda/memberLayer.zip'
      )
    });

    const lambdaFunction = new lambda.Function(this, 'Function', {
      functionName: RunbookFactory.getLambdaFunctionName(props.solutionId),
      handler: 'updatableRunbookProvider.lambda_handler',
      runtime: props.runtimePython,
      description: 'Custom resource to manage versioned SSM documents',
      code: lambda.Code.fromBucket(
          SolutionsBucket,
          props.solutionTMN + '/' + props.solutionVersion + '/lambda/updatableRunbookProvider.py.zip'
      ),
      environment: {
        LOG_LEVEL: 'info',
        SOLUTION_ID: `AwsSolution/${props.solutionId}/${props.solutionVersion}`
      },
      memorySize: 256,
      timeout: cdk.Duration.seconds(600),
      role: role,
      layers: [memberLambdaLayer],
      reservedConcurrentExecutions: 1
    });

    const cfnLambdaFunction = lambdaFunction.node.defaultChild as lambda.CfnFunction;
    cfnLambdaFunction.cfnOptions.metadata = {
      cfn_nag: {
        rules_to_suppress: [
          {
            id: 'W58',
            reason: 'False positive. Access is provided via a policy'
          },
          {
            id: 'W89',
            reason: 'There is no need to run this lambda in a VPC'
          }
        ]
      }
    };
  }

  static getLambdaFunctionName(solutionId: string): string {
    const RESOURCE_PREFIX = solutionId.replace(/^DEV-/, '');
    return `${RESOURCE_PREFIX}-SHARR-updatableRunbookProvider`;
  }

  static getServiceToken(scope: cdk.Construct, solutionId: string): string {
    const stack = cdk.Stack.of(scope);
    return `arn:${stack.partition}:lambda:${stack.region}:${stack.account}:function:${RunbookFactory.getLambdaFunctionName(solutionId)}`;
  }

  static getResourceType(): string {
    return 'Custom::UpdatableRunbook';
  }

  static createControlRunbook(scope: cdk.Construct, id: string, props: IssmPlaybookProps): cdk.CustomResource {
    let scriptPath = '';
    if (props.scriptPath == undefined ) {
      scriptPath = `${props.ssmDocPath}/scripts`;
    } else {
      scriptPath = props.scriptPath;
    }

    let commonScripts = '';
    if (props.commonScripts == undefined ) {
      commonScripts = '../common';
    } else {
      commonScripts = props.commonScripts;
    }

    const enableParam = new cdk.CfnParameter(scope, 'Enable ' + props.controlId, {
      type: 'String',
      description: `Enable/disable availability of remediation for ${props.securityStandard} version ${props.securityStandardVersion} Control ${props.controlId} in Security Hub Console Custom Actions. If NOT Available the remediation cannot be triggered from the Security Hub console in the Security Hub Admin account.`,
      default: 'Available',
      allowedValues: ['Available', 'NOT Available']
    });

    const installSsmDoc = new cdk.CfnCondition(scope, 'Enable ' + props.controlId + ' Condition', {
      expression: cdk.Fn.conditionEquals(enableParam, 'Available')
    });

    const ssmDocName = `SHARR-${props.securityStandard}_${props.securityStandardVersion}_${props.controlId}`;
    const ssmDocFQFileName = `${props.ssmDocPath}/${props.ssmDocFileName}`;
    const ssmDocType = props.ssmDocFileName.substring(props.ssmDocFileName.length - 4).toLowerCase();

    const ssmDocIn = fs.readFileSync(ssmDocFQFileName, 'utf8');

    let ssmDocOut: string = '';
    const re = /^(?<padding>\s+)%%SCRIPT=(?<script>.*)%%/;

    for (const line of ssmDocIn.split('\n')) {
      const foundMatch = re.exec(line);
      if (foundMatch && foundMatch.groups && foundMatch.groups.script) {
        let pathAndFileToInsert = foundMatch.groups.script;
        // If a relative path is provided then use it
        if (pathAndFileToInsert.substring(0, 7) === 'common/') {
          pathAndFileToInsert = `${commonScripts}/${pathAndFileToInsert.substring(7)}`;
        } else {
          pathAndFileToInsert = `${scriptPath}/${pathAndFileToInsert}`;
        }
        const scriptIn = fs.readFileSync(pathAndFileToInsert, 'utf8');
        for (const scriptLine of scriptIn.split('\n')) {
          ssmDocOut += foundMatch.groups.padding + scriptLine + '\n';
        }
      } else {
        ssmDocOut += line + '\n';
      }
    }

    const ssmDoc = new cdk.CustomResource(scope, id, {
      serviceToken: RunbookFactory.getServiceToken(scope, props.solutionId),
      resourceType: RunbookFactory.getResourceType(),
      properties: {
        Name: ssmDocName,
        Content: ssmDocOut,
        DocumentFormat: ssmDocType.toUpperCase(),
        VersionName: props.solutionVersion,
        DocumentType: 'Automation'
      }
    });

    const ssmDocCfnResource = ssmDoc.node.defaultChild as cdk.CfnCustomResource;
    ssmDocCfnResource.cfnOptions.condition = installSsmDoc;

    return ssmDoc;
  }

  static createRemediationRunbook(scope: cdk.Construct, id: string, props: RemediationRunbookProps) {
    const ssmDocName = `SHARR-${props.ssmDocName}`;
    let scriptPath = '';
    if (props.scriptPath == undefined) {
      scriptPath = 'ssmdocs/scripts';
    } else {
      scriptPath = props.scriptPath;
    }

    const ssmDocFQFileName = `${props.ssmDocPath}/${props.ssmDocFileName}`;
    const ssmDocType = props.ssmDocFileName.substring(props.ssmDocFileName.length - 4).toLowerCase();

    const ssmDocIn = fs.readFileSync(ssmDocFQFileName, 'utf8');

    let ssmDocOut: string = '';
    const re = /^(?<padding>\s+)%%SCRIPT=(?<script>.*)%%/;

    for (const line of ssmDocIn.split('\n')) {
      const foundMatch = re.exec(line);
      if (foundMatch && foundMatch.groups && foundMatch.groups.script) {
        const scriptIn = fs.readFileSync(`${scriptPath}/${foundMatch.groups.script}`, 'utf8');
        for (const scriptLine of scriptIn.split('\n')) {
          ssmDocOut += foundMatch.groups.padding + scriptLine + '\n';
        }
      } else {
        ssmDocOut += line + '\n';
      }
    }

    const runbook = new cdk.CustomResource(scope, id, {
      serviceToken: RunbookFactory.getServiceToken(scope, props.solutionId),
      resourceType: RunbookFactory.getResourceType(),
      properties: {
        Name: ssmDocName,
        Content: ssmDocOut,
        DocumentFormat: ssmDocType.toUpperCase(),
        DocumentType: 'Automation'
      }
    });

    return runbook;
  }
};
