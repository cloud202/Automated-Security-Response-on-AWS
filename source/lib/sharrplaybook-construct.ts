#!/usr/bin/env node
/*****************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

//
// Primary Stack - launched by Service Catalog in the Security Hub Admin account
// Creates CWE rules and custom actions, orchestrator step function
// Orchestrator lambdas are common to all Playbooks and deployed in the main stack
//
import * as cdk from '@aws-cdk/core';
import { StringParameter } from '@aws-cdk/aws-ssm';
import { Trigger, SsmPlaybook } from './ssmplaybook';
import { AdminAccountParm } from './admin_account_parm-construct';
import { RunbookFactory } from '../solution_deploy/lib/runbook_factory';

export interface IControl {
    control: string;
    executes?: string;
}
export interface PlaybookProps {
    description: string;
    solutionId: string;
    solutionVersion: string;
    solutionDistBucket: string;
    solutionDistName: string;
    remediations: IControl[];
    securityStandard: string;
    securityStandardLongName: string;
    securityStandardVersion: string;
}

export class PlaybookPrimaryStack extends cdk.Stack {

  constructor(scope: cdk.Construct, id: string, props: PlaybookProps) {
    super(scope, id, props);

    const stack = cdk.Stack.of(this)
    const RESOURCE_PREFIX = props.solutionId.replace(/^DEV-/,''); // prefix on every resource name
    const orchestratorArn = StringParameter.valueForStringParameter(this, `/Solutions/${RESOURCE_PREFIX}/OrchestratorArn`)

    // Register the playbook. These parameters enable the step function to route matching events
    new StringParameter(this, 'StandardShortName', {
        description: 'Provides a short (1-12) character abbreviation for the standard.',
        parameterName: `/Solutions/${RESOURCE_PREFIX}/${props.securityStandardLongName}/shortname`,
        stringValue: props.securityStandard
    });
    new StringParameter(this, 'StandardVersion', {
        description: 'This parameter controls whether the SHARR step function will process findings for this version of the standard.',
        parameterName: `/Solutions/${RESOURCE_PREFIX}/${props.securityStandardLongName}/${props.securityStandardVersion}/status`,
        stringValue: 'enabled'
    });

    new cdk.CfnMapping(this, 'SourceCode', {
        mapping: { "General": {
            "S3Bucket": props.solutionDistBucket,
            "KeyPrefix": props.solutionDistName + '/' + props.solutionVersion
        } }
    })

    const processRemediation = function(controlSpec: IControl): void {
        if ((controlSpec.executes != undefined) &&
            (controlSpec.control != controlSpec.executes)) {
            // This control is remapped to another
            new StringParameter(stack, `Remap ${props.securityStandard} ${controlSpec.control}`, {
                description: `Remap the ${props.securityStandard} ${controlSpec.control} finding to ${props.securityStandard} ${controlSpec.executes} remediation`,
                parameterName: `/Solutions/${RESOURCE_PREFIX}/${props.securityStandard}/${props.securityStandardVersion}/${controlSpec.control}/remap`,
                stringValue: `${controlSpec.executes}`
            });
        }
        let generatorId = ''
        if (props.securityStandard === 'CIS' && props.securityStandardVersion === '1.2.0') {
            // CIS 1.2.0 uses an arn-like format: arn:aws:securityhub:::ruleset/cis-aws-foundations-benchmark/v/1.2.0/rule/1.3
            generatorId = `arn:${stack.partition}:securityhub:::ruleset/${props.securityStandardLongName}/v/${props.securityStandardVersion}/rule/${controlSpec.control}`
        }
        else {
            generatorId = `${props.securityStandardLongName}/v/${props.securityStandardVersion}/${controlSpec.control}`
        }
        new Trigger(stack, `${props.securityStandard} ${controlSpec.control}`, {
            securityStandard: props.securityStandard,
            controlId: controlSpec.control,
            generatorId: generatorId,
            targetArn: orchestratorArn
        })
    }

    props.remediations.forEach(processRemediation)

  }
}

export interface MemberStackProps {
  description: string;
  solutionId: string;
  solutionVersion: string;
  solutionDistBucket: string;
  securityStandard: string;
  securityStandardVersion: string;
  securityStandardLongName: string;
  ssmdocs?: string;
  commonScripts?: string;
  remediations: IControl[];
}

export class PlaybookMemberStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props: MemberStackProps) {
    super(scope, id, props);
    const stack = cdk.Stack.of(this);

    let ssmdocs = '';
    if (props.ssmdocs == undefined) {
      ssmdocs = './ssmdocs';
    } else {
      ssmdocs = props.ssmdocs;
    }

    new AdminAccountParm(this, 'AdminAccountParameter', {
      solutionId: props.solutionId
    });

    const processRemediation = function(controlSpec: IControl): void {
      // Create the ssm automation document only if this is not a remapped control
      if (!(controlSpec.executes && controlSpec.control != controlSpec.executes)) {
        RunbookFactory.createControlRunbook(stack, `${props.securityStandard} ${controlSpec.control}`, {
          securityStandard: props.securityStandard,
          securityStandardVersion: props.securityStandardVersion,
          controlId: controlSpec.control,
          ssmDocPath: ssmdocs,
          ssmDocFileName: `${props.securityStandard}_${controlSpec.control}.yaml`,
          solutionVersion: props.solutionVersion,
          solutionDistBucket: props.solutionDistBucket,
          solutionId: props.solutionId,
          commonScripts: props.commonScripts
        });
      }
    };

    props.remediations.forEach(processRemediation);
  }
}
