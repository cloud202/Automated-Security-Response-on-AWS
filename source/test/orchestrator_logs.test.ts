/*****************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

 import { SynthUtils } from '@aws-cdk/assert';
 import * as cdk from '@aws-cdk/core';
 import { OrchLogStack } from '../solution_deploy/lib/orchestrator-log-stack';
 import { AwsSolutionsChecks } from 'cdk-nag'
import { Aspects } from '@aws-cdk/core'
 
 const app = new cdk.App();
 
 function getTestStack(): cdk.Stack {
   const app = new cdk.App();
   const stack = new OrchLogStack(app, 'roles', {
     description: 'test;',
     solutionId: 'SO0111',
     logGroupName: 'TestLogGroup'
   })
   Aspects.of(app).add(new AwsSolutionsChecks({verbose: true}))
   return stack;
 }
 test('Global Roles Stack', () => {
   expect(SynthUtils.toCloudFormation(getTestStack())).toMatchSnapshot();
 });
 