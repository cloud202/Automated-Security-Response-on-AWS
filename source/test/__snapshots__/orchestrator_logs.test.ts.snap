// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global Roles Stack 1`] = `
Object {
  "Conditions": Object {
    "EncryptedLogGroup": Object {
      "Fn::And": Array [
        Object {
          "Condition": "isNotGovCloud",
        },
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "ReuseOrchestratorLogGroup",
            },
            "no",
          ],
        },
      ],
    },
    "UnencryptedLogGroup": Object {
      "Fn::And": Array [
        Object {
          "Fn::Not": Array [
            Object {
              "Condition": "isNotGovCloud",
            },
          ],
        },
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "ReuseOrchestratorLogGroup",
            },
            "no",
          ],
        },
      ],
    },
    "isNotGovCloud": Object {
      "Fn::Not": Array [
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Partition",
            },
            "aws-us-gov",
          ],
        },
      ],
    },
  },
  "Description": "test;",
  "Parameters": Object {
    "KmsKeyArn": Object {
      "Description": "ARN of the KMS key to use to encrypt log data.",
      "Type": "String",
    },
    "ReuseOrchestratorLogGroup": Object {
      "AllowedValues": Array [
        "yes",
        "no",
      ],
      "Default": "no",
      "Description": "Reuse existing Orchestrator Log Group? Choose \\"yes\\" if the log group already exists, else \\"no\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "OrchestratorLogsEFDFFA92": Object {
      "Condition": "UnencryptedLogGroup",
      "DeletionPolicy": "Retain",
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W84",
              "reason": "KmsKeyId is not supported in GovCloud.",
            },
          ],
        },
      },
      "Properties": Object {
        "LogGroupName": "TestLogGroup",
        "RetentionInDays": 365,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "OrchestratorLogsEncrypted072D6E38": Object {
      "Condition": "EncryptedLogGroup",
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KmsKeyId": Object {
          "Ref": "KmsKeyArn",
        },
        "LogGroupName": "TestLogGroup",
        "RetentionInDays": 365,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
